### PR TITLE
Story/CCP-203

### PR DESCRIPTION
**Pull request by mkarthik90**

### CODE REVIEW

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/rest/Concepts.java [hunk 1]
```diff
@@ -115,13 +115,19 @@
...
         }
 
+        JsonValidationResult result = checkJsonObject(jsonObject);
+
+        if (result.isValid() && jsonObject.get(JsonFields.WORDNET_ID) != null) {
+            result = checkJsonObjectForWrapper(jsonObject);
+        }
+
+        if (!result.isValid())
+            return new ResponseEntity<String>(result.getMessage(), HttpStatus.BAD_REQUEST);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/rest/Concepts.java]: parentheses missing

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/rest/Concepts.java [hunk 2]
```diff
@@ -184,14 +190,21 @@
 
             JsonValidationResult result = null;
 
+            result = checkJsonObject(jsonObject);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/rest/Concepts.java]: merge with line 191

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptLookUpTest.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        Mockito.when(typeManager.getType(typeId2)).thenReturn(type2);
+        Mockito.when(messageFactory.getMessageFactory(MediaType.APPLICATION_XML_VALUE)).thenReturn(xmlMEssageFactory);
+        Mockito.when(messageFactory.getMessageFactory(MediaType.APPLICATION_JSON_VALUE)).thenReturn(jsonMessageFactory);
+
+        Mockito.when(xmlMEssageFactory.createConceptMessage()).thenReturn(new XMLConceptMessage(uriCreator));
+        Mockito.when(jsonMessageFactory.createConceptMessage()).thenReturn(new JsonConceptMessage(uriCreator));
+    }
+
+    @Test
+    public void getWordNetEntryTest()
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptLookUpTest.java]: please follow our naming conventions for tests:
https://diging.atlassian.net/wiki/display/DIGING/Unit+Tests

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptLookUpTest.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+        ConceptEntry entry2 = new ConceptEntry();
+        entry2.setId("WID-456");
+        entry2.setWord("robert-1");
+        entry1.setPos(POS.NOUN);
+
+        entries[0] = entry1;
+        entries[1] = entry2;
+
+        Mockito.when(dictManager.getConceptListEntriesForWordPOS("robert", POS.NOUN, null)).thenReturn(entries);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptLookUpTest.java]: these strings should be in constants or declared at the beginning of the method to make maintenance easier.

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptLookUpTest.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+import edu.asu.conceptpower.core.ConceptEntry;
+import edu.asu.conceptpower.core.ConceptType;
+import edu.asu.conceptpower.rest.ConceptLookup;
+import edu.asu.conceptpower.rest.msg.IMessageConverter;
+import edu.asu.conceptpower.rest.msg.IMessageRegistry;
+import edu.asu.conceptpower.rest.msg.json.JsonConceptMessage;
+import edu.asu.conceptpower.rest.msg.xml.XMLConceptMessage;
+import junit.framework.Assert;
+
+public class ConceptLookUpTest {
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptLookUpTest.java]: the original class is spelled slightly different

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptLookUpTest.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        entries[1] = entry2;
+
+        Mockito.when(dictManager.getConceptListEntriesForWordPOS("robert", POS.NOUN, null)).thenReturn(entries);
+        ResponseEntity<String> response = conceptLookup.getWordNetEntryInRdf("robert", POS.NOUN);
+        Mockito.verify(rdfFactory).generateRDF(entries);
+        Assert.assertEquals(HttpStatus.OK, response.getStatusCode());
+
+    }
+
+}
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptLookUpTest.java]: your tests don't seem to test the most important part, which is that if I search for "pony" it actually returns the concepts containing "pony" (or "robert") and the correct number.

Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptLookUpTest.java]: What about searches that don't yield any results?

Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptLookUpTest.java]: what about searches with special characters?

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptSearchTest.java [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+
+    @Mock
+    private URIHelper uriCreator;
+
+    @Mock(name = "jsonMessageFactory")
+    private IMessageConverter jsonMessageFactory;
+
+    private ConceptSearchParameters conceptSearchParameter;
+
+    private String xmlNoRecordsFoundError = "<conceptpowerReply xmlns:digitalHPS=\"http://www.digitalhps.org/\"><digitalHPS:error error_message=\"No records found for the search condition.\" ></digitalHPS:error></conceptpowerReply>";
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptSearchTest.java]: should be in external file

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptSearchTest.java [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+    private int numberOfRecordsPerPage = 20;
+    private String jsonNoRecordsFoundError = "{\"message\":\"No records found for the search condition.\"}";
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptSearchTest.java]: ideally, this one too

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptSearchTest.java [hunk 9]
```diff
@@ -0,0 +1,261 @@
+package edu.asu.conceptpower.servlet.rest;
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptSearchTest.java]: adhere to naming conventions

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptSearchTest.java [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+                .thenReturn(new JsonConceptMessage(uriCreator));
+        conceptSearchParameter = new ConceptSearchParameters();
+        conceptSearchParameter.setPos("noun");
+        conceptSearchParameter.setWord("pony");
+        conceptSearchParameter.setConcept_list("Test-List");
+        conceptSearchParameter.setType_id("TYPE-123");
+        conceptSearchParameter.setOperator(SearchParamters.OP_AND);
+        conceptSearchParameter.setNumber_of_records_per_page(numberOfRecordsPerPage);
+
+        String typeId = "TYPE-123";
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptSearchTest.java]: all those strings should be in constants or final fields at the beginning of the method

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptSearchTest.java [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+
+        Mockito.verify(manager).searchForConceptByPageNumberAndFieldMap(Mockito.anyMapOf(String.class, String.class),
+                Mockito.anyString(), Mockito.eq(1), Mockito.eq(numberOfRecordsPerPage));
+
+        Assert.assertEquals(HttpStatus.OK, output.getStatusCode());
+        Assert.assertEquals(xmlNoRecordsFoundError, output.getBody());
+    }
+
+    @Test
+    public void testForPOSError()
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptSearchTest.java]: what about a POS value that doesn't exists (e.g. "subject" or "bla")

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptsTest.java [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+    @Before
+    public void init() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void addConceptWithInvalidTypeTest()
+            throws IllegalAccessException, LuceneException, IndexerRunningException {
+
+        String input = " { \"word\": \"kitty\",  \"pos\": \"noun\", \"conceptlist\": \"mylist\", \"description\": \"Soft kitty, sleepy kitty, little ball of fur.\",\"type\": \"3b755111-545a-4c1c-929c-a2c0d4c3913b\"}";
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptsTest.java]: externalize

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/SynonymSearchTest.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        conceptType.setCreatorId("admin");
+        conceptType.setDescription("Concept Type");
+
+        Mockito.when(dictManager.getSynonymsForConcept(synonymId)).thenReturn(entries);
+        Mockito.when(typeManager.getType("TYPE-123")).thenReturn(conceptType);
+
+    }
+
+    @Test
+    public void getSynonymsForInvalidId() throws JsonProcessingException {
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/SynonymSearchTest.java]: what about an id that doesn't exist?

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/TypeIdLookupTest.java [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+        HttpServletRequest req = Mockito.mock(HttpServletRequest.class);
+        Mockito.when(req.getParameter("id"))
+                .thenReturn("http://www.digitalhps.org/types/TYPE_0c2a2a8c-c6d8-4d25-8f07-689fd8c27b0b");
+
+        ResponseEntity<String> typeMessage = typeIdLookup.getTypeById(req, MediaType.APPLICATION_JSON_VALUE);
+
+        RestTestUtility.testValidJson(typeMessage.getBody());
+        Assert.assertEquals(HttpStatus.OK, typeMessage.getStatusCode());
+
+    }
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/TypeIdLookupTest.java]: as before, what about checking that the actual content is correct?

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/TypeIdLookupTest.java [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+
+    @Test
+    public void testForTypeNotFound() throws JsonProcessingException {
+        String typeId = "0c2a2a8c-c6d8-4d25-8f07-689fd8c27b0w";
+        Mockito.when(typeManager.getType(typeId)).thenReturn(null);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/TypeIdLookupTest.java]: again, what about non-existing id?

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptLookUpTest.java (hunk 6 above)
Comment (mkarthik90) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptLookUpTest.java]: I am not able to understand.

Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/servlet/rest/ConceptLookUpTest.java]: well, does the search for pony actually return the results for pony (maybe it replaces all words with "blablabla')

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/rest/SynonymSearchTest.java [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+
+    @Test
+    public void test_getSynonymsForId_nonExistingId() throws JsonProcessingException, LuceneException {
+        HttpServletRequest req = Mockito.mock(HttpServletRequest.class);
+        final String wordnet = "WID-PONY-123";
+        Mockito.when(req.getParameter("id")).thenReturn("http://www.digitalhps.org/concepts/" + wordnet);
+        Mockito.when(dictManager.getSynonymsForConcept(wordnet))
+                .thenReturn(Collections.<ConceptEntry> emptyList().toArray(new ConceptEntry[0]));
+        ResponseEntity<String> output = synonymSearch.getSynonymsForId(req, MediaType.APPLICATION_XML_VALUE);
+        Assert.assertEquals(HttpStatus.OK, output.getStatusCode());
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/rest/SynonymSearchTest.java]: what would be the expected output here?

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/rest/ConceptSearchTest.java [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+        Map<String, String> map = new HashMap<>();
+        map.put("pos", pos);
+        map.put("type-id", typeId);
+        map.put("concept_list", conceptListName);
+        map.put("word", word);
+
+        int numberOfResults = 100;
+
+        Mockito.when(manager.getTotalNumberOfRecordsForSearch(Mockito.anyMapOf(String.class, String.class),
+                Mockito.anyString())).thenReturn(numberOfResults);
```
Comment (mkarthik90) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/rest/ConceptSearchTest.java]: I changed this one to 2. And when there are no records found I made this one to 0

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/rest/ConceptLookupTest.java [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+        Mockito.when(typeManager.getType(typeId2)).thenReturn(type2);
+        Mockito.when(messageFactory.getMessageFactory(MediaType.APPLICATION_XML_VALUE)).thenReturn(xmlMEssageFactory);
+        Mockito.when(messageFactory.getMessageFactory(MediaType.APPLICATION_JSON_VALUE)).thenReturn(jsonMessageFactory);
+
+        Mockito.when(xmlMEssageFactory.createConceptMessage()).thenReturn(new XMLConceptMessage(uriCreator));
+        Mockito.when(jsonMessageFactory.createConceptMessage()).thenReturn(new JsonConceptMessage(uriCreator));
+    }
+
+    @Test
+    public void test_getWordNetEntry_successINXml()
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/rest/ConceptLookupTest.java]: this still seems to be the same?

**Comment by jdamerow** [comment]
ok to test

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/rest/ConceptLookupTest.java (hunk 18 above)
Comment (mkarthik90) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/rest/ConceptLookupTest.java]: @jdamerow Sorry fixed it now.

**Pull request merged**